smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpProtocolGeneratorUtils.java

```diff
@@ -238,4 +242,38 @@ static void generateCollectBodyString(GenerationContext context) {
 
         return errorShapes;
     }
+
+    /**
+     * Writes resolved hostname, prepending existing hostname with hostPrefix and replacing each hostLabel with
+     * the corresponding top-level input member value.
+     *
+     * @param context The generation context.
+     * @param operation The operation to generate for.
+     */
+    static void writeHostPrefix(
+            GenerationContext context,
+            OperationShape operation
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        EndpointTrait trait = operation.getTrait(EndpointTrait.class).get();
+        writer.write("let resolvedHostname = (context.endpoint as any).hostname;");
+        // Check if disableHostPrefixInjection has been set to true at runtime
+        writer.openBlock("if (context.disableHostPrefix !== true) {", "}", () -> {
+            writer.write("resolvedHostname = $S + resolvedHostname;", trait.getHostPrefix().toString());
+            List<Pattern.Segment> prefixLabels = trait.getHostPrefix().getLabels();
+            Shape inputShape = context.getModel().expectShape(operation.getInput().get());
+            for (Pattern.Segment label : prefixLabels) {
+                MemberShape target = inputShape.members().stream().filter(s -> s.getMemberName().equals(
```

Comment:
Nit: the member is not the target, it's a member. A member actually has a target shape, so this terminology is confusing. Maybe `labelTarget`? Or just `member`?